File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/widget/EmailWebView.java
/*
 * © 2016-2018 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.ui.widget;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.support.customtabs.CustomTabsIntent;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.flowcrypt.email.R;
import com.flowcrypt.email.model.MessageEncryptionType;
import com.flowcrypt.email.ui.activity.CreateMessageActivity;

/**
 * The custom realization of {@link WebView}
 *
 * @author Denis Bondarenko
 *         Date: 02.09.2017
 *         Time: 12:19
 *         E-mail: DenBond7@gmail.com
 */

public class EmailWebView extends WebView {
    public EmailWebView(Context context) {
        super(context);
    }

    public EmailWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public EmailWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    /**
     * This method does job of configure the current {@link WebView}
     */
    public void configure() {
        setVerticalScrollBarEnabled(false);
        setScrollBarStyle(SCROLLBARS_INSIDE_OVERLAY);
        setOverScrollMode(OVER_SCROLL_NEVER);
        setWebViewClient(new CustomWebClient(getContext()));

        WebSettings webSettings = this.getSettings();

        webSettings.setUseWideViewPort(true);
        webSettings.setLoadWithOverviewMode(true);
        webSettings.setSupportZoom(true);
        webSettings.setBuiltInZoomControls(true);
        webSettings.setDisplayZoomControls(false);
        webSettings.setLoadsImagesAutomatically(true);
        webSettings.setLoadsImagesAutomatically(true);
        webSettings.setJavaScriptEnabled(false);
        webSettings.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.SINGLE_COLUMN);
    }

    /**
     * The custom realization of {@link WebViewClient}
     */
    private static class CustomWebClient extends WebViewClient {
        private Context context;

        CustomWebClient(Context context) {
            this.context = context;
        }

        @SuppressWarnings("deprecation")
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            if (url.startsWith(SCHEME_MAILTO)) {
                handleEmailLinks(Uri.parse(url));
                return false;
            } else {
                showUrlUsingChromeCustomTabs(Uri.parse(url));
                return true;
            }
        }

        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            if (request.getUrl().getScheme().equalsIgnoreCase("mailto")) {
                handleEmailLinks(request.getUrl());
                return true;
            } else {
                showUrlUsingChromeCustomTabs(request.getUrl());
                return true;
            }
        }

        /**
         * Handle email links and open the internal compose screen.
         *
         * @param uri {@link Uri} with mailto: scheme.
         */
        private void handleEmailLinks(Uri uri) {
            Intent intent = CreateMessageActivity.generateIntent(context, null, MessageEncryptionType.ENCRYPTED);
            intent.setAction(Intent.ACTION_SENDTO);
            intent.setData(uri);
            context.startActivity(intent);
        }


        /**
         * Use {@link CustomTabsIntent} to show some url.
         *
         * @param uri The {@link Uri} which contains a url.
         */
        private void showUrlUsingChromeCustomTabs(Uri uri) {
            CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
            CustomTabsIntent customTabsIntent = builder.build();
            builder.setToolbarColor(ContextCompat.getColor(context, R.color.colorPrimary));
            customTabsIntent.launchUrl(context, uri);
        }
    }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/ui/widget/EmailWebView.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: denbond7
 */

package com.flowcrypt.email.ui.widget

import android.content.Context
import android.content.Intent
import android.graphics.Point
import android.net.Uri
import android.util.AttributeSet
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.core.net.toUri
import com.flowcrypt.email.extensions.android.webkit.setupDayNight
import com.flowcrypt.email.extensions.android.webkit.showUrlUsingChromeCustomTabs
import com.flowcrypt.email.model.MessageType
import com.flowcrypt.email.ui.activity.CreateMessageActivity

/**
 * The custom realization of [WebView]
 *
 * @author Denys Bondarenko
 */
class EmailWebView : WebView {
  private var onPageLoadingListener: OnPageLoadingListener? = null

  constructor(context: Context) : super(context)

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

  constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
    context,
    attrs,
    defStyleAttr
  )

  val isContentExpandedAfterInitialLoading: Boolean
    get() = currentSize.y > sizeOfContentAfterLoading.y

  private var sizeOfContentAfterLoading = Point()
  private var currentSize = Point()

  private var isContentLoaded = false

  override fun loadDataWithBaseURL(
    baseUrl: String?,
    data: String,
    mimeType: String?,
    encoding: String?,
    historyUrl: String?
  ) {
    isContentLoaded = false
    super.loadDataWithBaseURL(baseUrl, data, mimeType, encoding, historyUrl)
  }

  override fun onSizeChanged(w: Int, h: Int, ow: Int, oh: Int) {
    super.onSizeChanged(w, h, ow, oh)
    currentSize = Point(w, h)
    if (isContentLoaded && sizeOfContentAfterLoading.x == 0) {
      sizeOfContentAfterLoading = Point(w, h)
    }
  }

  /**
   * This method does job of configure the current [WebView]
   */
  fun configure() {
    isVerticalScrollBarEnabled = false
    scrollBarStyle = SCROLLBARS_INSIDE_OVERLAY
    overScrollMode = OVER_SCROLL_NEVER
    webViewClient = CustomWebClient(context)
    webChromeClient = object : WebChromeClient() {
      override fun onProgressChanged(view: WebView, newProgress: Int) {
        super.onProgressChanged(view, newProgress)
        isContentLoaded = newProgress >= 100
        onPageLoadingListener?.onPageLoading(newProgress)
      }
    }

    val webSettings = this.settings

    webSettings.useWideViewPort = true
    webSettings.loadWithOverviewMode = true
    webSettings.setSupportZoom(true)
    webSettings.builtInZoomControls = true
    webSettings.displayZoomControls = false
    webSettings.loadsImagesAutomatically = true
    webSettings.blockNetworkLoads = true
    webSettings.javaScriptEnabled = false
    webSettings.allowFileAccess = false

    setupDayNight()
  }

  fun setOnPageLoadingListener(onPageLoadingListener: OnPageLoadingListener) {
    this.onPageLoadingListener = onPageLoadingListener
  }

  interface OnPageLoadingListener {
    fun onPageLoading(newProgress: Int)
  }

  /**
   * The custom realization of [WebViewClient]
   */
  private class CustomWebClient(private val context: Context) :
    WebViewClient() {

    @Deprecated("Deprecated in Java")
    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
      return if (url.startsWith(SCHEME_MAILTO)) {
        handleEmailLinks(url.toUri())
        false
      } else {
        showUrlUsingChromeCustomTabs(context = context, uri = url.toUri())
        true
      }
    }

    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
      if ("mailto".equals(request.url.scheme, ignoreCase = true)) {
        handleEmailLinks(request.url)
      } else {
        showUrlUsingChromeCustomTabs(context = context, uri = request.url)
      }

      return true
    }

    /**
     * Handle email links and open the internal compose screen.
     *
     * @param uri [Uri] with mailto: scheme.
     */
    private fun handleEmailLinks(uri: Uri) {
      val intent = CreateMessageActivity.generateIntent(context, MessageType.NEW)
      intent.action = Intent.ACTION_SENDTO
      intent.data = uri
      context.startActivity(intent)
    }
  }
}
